src/control/server/instance_storage_rpc.go

```diff
@@ -163,6 +166,63 @@ func (ei *EngineInstance) StorageFormatSCM(ctx context.Context, force bool) (mRe
 	return
 }
 
+func setLnkStats(log logging.TraceLogger, inStr string, health *ctlpb.BioHealthResp) error {
```

Comment:
you want to send the protobuf struct into a method in the hardware library to get populated? that doesn't really sound like a good fit to me

Reply:
The use of reflect seems appropriate here as it can be used to avoid specifying each health stat explicitly in a switch case when scanning through the input text (lspci output).

Reply:
I've attempt to address your comments by moving to the hardware library and commenting on the use of reflect.

Reply:
> I've attempt to address your comments by moving to the hardware library and commenting on the use of reflect.

OK, I think I'm not communicating effectively. I am saying that the approach you've got here, even after moving the function into the hardware library, is very narrowly focused on this specific use case of decorating a protobuf message in a gRPC handler before returning it to dmg.

The code works now for this specific use case, sure. What about down the road when we have a different use case where we'd like to key some decisions based off of the link speed or link status? How would we reuse this code you're adding in order to accomplish that? As far as I can tell, it would be pretty difficult. The developer would have to know about this new `RunLspciWithInput()` function and figure out how to use it. It takes a string and returns a string, so they'd have to figure out what those look like and how to generate/interpret them.

And then we've got this `SetPciLinkStats()` function which accepts a string and a protobuf message. In order to use this, a developer has to figure out what the input string looks like or where it comes from. At the moment, I guess you'd have to copy whatever is going on in `addLinkInfoToHealthStats()`? Why should another developer (or you, after a few months of looking at other stuff) have to go through and figure all of this out? Wouldn't it be nicer to just call into the hardware library and get a list of [PCIDevice](https://github.com/daos-stack/daos/blob/master/src/control/lib/hardware/pci.go#L438)s with all of this info populated on them?

The solution as it is now accomplishes the immediate task, but we really should be thinking in terms of libraries and reuse. A change to the logic for getting PCI link speed should ideally not result in changes outside of the hardware library, if we've done it correctly. It's more work up-front, but it yields dividends down the road as the codebase grows and more developers are working with it.

I'm willing to be convinced that what I'm asking for is impossible or unreasonable, but I don't think it is.

Reply:
Thanks for the detailed explanation I will try again.

Reply:
By implementing the approach you are prescribing we lose the brevity and simplicity benefits of the current solution (using reflect to directly parse lspci into BioHealthResp). Forcing delegation into the hardware module (obviously there is bias for using a module one has authored) may be cleaner from one perspective but causes unnecessary bloat in this solution IMO and indirection (passing config space string received over dRPC, decoding individual parameters from an already decoded and formatted lspci output, populating PCIDevice structs, copying fields to a BioHealthResp struct) and the promise of future benefits doesn't warrant the level of change requested. In my mind this is an example of premature optimisation. However as this is the consensus I will disagree and commit...

Reply:
Having looked again there is at least one significant drawback to the suggested approach - decoding the config space bytes through lspci then picking apart the text generated to convert the "humanised" metrics to raw types is nonsensical and prone to error if lspci code changes (unlikely) or some output format is not handled correctly by us. The ideal solution would be to provide a go implementation of PCIe config space decoder (or implement on the engine side which I did take a look into) but this to me is not within the scope of effort to address the minor customer ask. I would like to ask that we accept the current changes and look to spend more effort on a fully fledged PCIDevice implementation when more feature development work is required.

Reply:
Frankly, I'm not convinced that the approach of shelling out to lspci (which may or may not be installed on a system (you haven't added it as a dependency in the RPM specfile or debian control file) is great to begin with. I played around a little with creating a binding for pciutils under `lib/hardware` and it wasn't too bad to work with. I don't have time to flesh that out myself, though.

I understand that you're just trying to quickly satisfy a customer request, but the proposed approach adds technical debt. Quick solutions here and there pile up into a bad codebase and ultimately slow down velocity because changes are hard to make when they cut across multiple layers and it's difficult to figure out how to use code that wasn't designed for testability and reuse.

Reply:
do you have a branch for the pciutils bindings?

Reply:
> do you have a branch for the pciutils bindings?

Extremely WIP to play with the idea: https://github.com/daos-stack/daos/pull/14521

Reply:
have integrated